### PR TITLE
feat: add server-side validation to useForm

### PR DESCRIPTION
**Pull request by alicanerdurmaz**
After triggering the `onMutationError` callback, `useForm` will automatically set the thrown errors to the corresponding form fields.
From now, when the `dataProvider` throws an `HttpError` with an `errors` field, the form will display the errors in the appropriate fields. With this feature, users can handle server-side validation errors out-of-the-box.

- Material UI and Chakra UI examples are added
- Mantine example is added
- Ant Design example is added

> [Documentation Link](https://refine.dev/docs/advanced-tutorials/forms/server-side-form-validation/)

### Test plan (required)

Tests are added for new features.

image

### Self Check before Merge

Please check all items below before review.

-   [x] Corresponding issues are created/updated or not needed
-   [x] Docs are updated/provided or not needed
-   [x] Examples are updated/provided or not needed
-   [x] TypeScript definitions are updated/provided or not needed
-   [x] Tests are updated/provided or not needed
-   [x] Changesets are provided or not needed

### CODE REVIEW

File: .changeset/gentle-crabs-cough.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"@refinedev/mantine": minor
+---
+
+feat: After the `onMutationError` callback is triggered, `useForm` will automatically set the thrown errors to the corresponding form fields.
```
Comment (BatuhanW) [inline .changeset/gentle-crabs-cough.md:5]:
```suggestion
feat: When `onMutationError` callback is triggered, `useForm` will automatically assign errors to the corresponding form fields.
```

Should we mention about `errors` field? We mention it in the next paragraph, looks like duplicated tho. Maybe we can merge 2 sentences.

File: .changeset/good-files-attend.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        | boolean
+        | { key: string; message: string };
+}
+
+export interface HttpError extends Record<string, any> {
+    message: string;
+    statusCode: number;
+    errors?: ValidationErrors;
+}
+```
```
Comment (BatuhanW) [inline .changeset/good-files-attend.md:22]: Should we add an example error with fields to demonstrate?

File: .changeset/new-suits-ring.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+---
+"@refinedev/antd": minor
+---
+
+feat: After the `onMutationError` callback is triggered, `useForm` will automatically set the thrown errors to the corresponding form fields.
+From know, when the `dataProvider` throws an `HttpError` with an `errors` field, the form will display the errors in the appropriate fields. With this feature users can handle server-side validation errors out-of-the-box.
```
Comment (BatuhanW) [inline .changeset/new-suits-ring.md:6]: Let's change this one too when we re-phrase changelog about mantine.

File: documentation/docs/advanced-tutorials/forms/partials/_mantine-server-side-form-validation.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    );
+};
+// visible-block-end
+
+setRefineProps({
+    dataProvider: {
+        ...dataProvider("https://api.fake-rest.refine.dev"),
+        update: async (resource: string, params: any) => {
+            return Promise.reject({
+                message: "Update is not supported in this example.",
```
Comment (BatuhanW) [inline documentation/docs/advanced-tutorials/forms/partials/_mantine-server-side-form-validation.md:59]: I think this error is misleading. We can use some generic message like `Error updating record` or something like that.

File: documentation/docs/advanced-tutorials/forms/server-side-form-validation.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+## Why Server-Side Validation?
+
+Client-side validation offers a responsive user experience by providing immediate feedback without server round trips. However, it should not be considered a substitute for server-side validation due to its vulnerability to bypassing. Server-side form validation is essential for ensuring data integrity, security, and consistency. It acts as an additional layer that complements client-side validation, preventing malicious or incorrect data from being processed. While client-side validation is valuable, it should not be relied upon exclusively, as server-side validation provides a more robust and reliable validation mechanism.
+
+## Server-Side Validation with **refine**
+
+**refine** supports server-side validation out of the box for all supported UI frameworks (Ant Design, Material UI, Mantine, Chakra UI).
+It requires that the `dataProvider` returning rejected promise with the following shape:
```
Comment (BatuhanW) [inline documentation/docs/advanced-tutorials/forms/server-side-form-validation.md:20]:
```suggestion
It expects `dataProvider` to return rejected promise with the following shape:
```

File: documentation/docs/advanced-tutorials/forms/server-side-form-validation.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+`errors` fields can be `string` or `string[]` or `boolean` or `{ key: string; message: string }`
+
+-   `string` or `string[]`: If the field is an array, multiple error messages will be displayed. If the field is a string, only one error message will be displayed.
+-   `boolean`: If the field is `true`, "This field is required." message will be displayed. If the field is `false`, no error message will be displayed.
+-   `{ key: string; message: string }`: If the field is an object, the `key` field will be used as a translation key. If the `key` is not found in the translation, the `message` field will be displayed.
+
+### How does it work?
+
+When the `dataProvider` returns rejected promise with [`errors`][http-error] field, the `useForm` hook will set the [`errors`][http-error] state with the `errors` returned from the `dataProvider`.
```
Comment (BatuhanW) [inline documentation/docs/advanced-tutorials/forms/server-side-form-validation.md:51]:
```suggestion
When the `dataProvider` returns a rejected promise with [`errors`][http-error] field, `useForm` hook will assign [`errors`][http-error] to the respective form fields.
```

File: documentation/docs/examples/form/antd/serverSideFormValidation.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+---
+id: serverSideFormValidation
+title: Server-Side Form Validation
+example-tags: [form, antd]
+---
+
+You can make server-side form validation out-of-the-box with [Ant Design useForm][antd-use-form].
```
Comment (BatuhanW) [inline documentation/docs/examples/form/antd/serverSideFormValidation.md:7]:
```suggestion
You can handle server-side form validation errors out-of-the-box with [Ant Design useForm][antd-use-form].
```

File: documentation/docs/examples/form/mantine/serverSideFormValidation.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
+---
+id: serverSideFormValidation
+title: Server-Side Form Validation
+example-tags: [form, mantine]
+---
+
+You can make server-side form validation out-of-the-box with [Mantine useForm][mantine-use-form].
```
Comment (BatuhanW) [inline documentation/docs/examples/form/mantine/serverSideFormValidation.md:7]: It gives an impressions that we are doing server-side validation, in fact, we are handling validation errors coming from the server side.

File: documentation/docs/api-reference/mantine/hooks/form/useForm.md [hunk 9]
```diff
@@ -1074,6 +1079,9 @@ const { overtime } = useForm();
 
 console.log(overtime.elapsedTime); // undefined, 1000, 2000, 3000 4000, ...
 ```
+
+# <<<<<<< HEAD
```
Comment (yildirayunlu) [inline documentation/docs/api-reference/mantine/hooks/form/useForm.md:1083]:
```suggestion
```

**Review by yildirayunlu — CHANGES_REQUESTED** [review]

File: examples/server-side-form-validation-antd/index.html [hunk 10]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <meta charset="utf-8" />
+        <link rel="icon" href="/favicon.ico" />
+        <meta name="viewport" content="width=device-width, initial-scale=1" />
+        <meta name="theme-color" content="#000000" />
+        <meta name="description" content="Web site created using refine" />
+        <link rel="apple-touch-icon" href="/logo192.png" />
```
Comment (salihozdemir) [inline examples/server-side-form-validation-antd/index.html:9]:
```suggestion
```

File: examples/server-side-form-validation-antd/public/manifest.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "refine base example",
+  "name": "refine base example",
```
Comment (salihozdemir) [inline examples/server-side-form-validation-antd/public/manifest.json:3]:
```suggestion
  "short_name": "refine server-side validation Ant Design example",
  "name": "refine server-side validation Ant Design example",
```

File: examples/server-side-form-validation-antd/src/pages/posts/create.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import { Form, Input, Select } from "antd";
+
+import MDEditor from "@uiw/react-md-editor";
+
+import { IPost, ICategory } from "../../interfaces";
+
+export const PostCreate: React.FC<IResourceComponentsProps> = () => {
+    const { formProps, saveButtonProps } = useForm<IPost>({
+        // warnWhenUnsavedChanges: true,
+    });
```
Comment (salihozdemir) [inline examples/server-side-form-validation-antd/src/pages/posts/create.tsx:14]:
```suggestion
    const { formProps, saveButtonProps } = useForm<IPost>();
```

File: examples/server-side-form-validation-antd/src/pages/posts/edit.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import { IPost, ICategory } from "../../interfaces";
+
+export const PostEdit: React.FC<IResourceComponentsProps> = () => {
+    const { formProps, saveButtonProps, queryResult } = useForm<
+        IPost,
+        HttpError,
+        IPost
+    >({
+        warnWhenUnsavedChanges: true,
+    });
```
Comment (salihozdemir) [inline examples/server-side-form-validation-antd/src/pages/posts/edit.tsx:18]:
```suggestion
    >();
```
By default, it is already `true`

File: examples/server-side-form-validation-material-ui/package.json [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    "@emotion/styled": "^11.8.1",
+    "@mui/lab": "^5.0.0-alpha.85",
+    "@mui/material": "^5.8.6",
+    "@mui/x-data-grid": "^6.6.0",
+    "react": "^18.0.0",
+    "react-dom": "^18.0.0",
+    "react-router-dom": "^6.8.1",
+    "@uiw/react-md-editor": "^3.19.5",
+    "react-scripts": "^5.0.0",
+    "cross-env": "^7.0.3",
```
Comment (yildirayunlu) [inline examples/server-side-form-validation-material-ui/package.json:22]:
```suggestion
```

File: examples/server-side-form-validation-chakra-ui/index.html [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+<html lang="en">
+    <head>
+        <meta charset="UTF-8" />
+        <link rel="icon" href="/src/favicon.ico" />
+        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+        <title>refine base chakra-ui example</title>
+        <link
+            href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;500;700;800&display=swap"
+            rel="stylesheet"
+        />
```
Comment (salihozdemir) [inline examples/server-side-form-validation-chakra-ui/index.html:11]:
```suggestion
        
```

**Review by yildirayunlu — CHANGES_REQUESTED** [review]

File: examples/server-side-form-validation-mantine/index.html [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+<html lang="en">
+    <head>
+        <meta charset="UTF-8" />
+        <link rel="icon" href="/src/favicon.ico" />
+        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+        <title>refine server side form validation mantine example</title>
+        <link
+            href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;500;700;800&display=swap"
+            rel="stylesheet"
+        />
```
Comment (salihozdemir) [inline examples/server-side-form-validation-mantine/index.html:11]:
```suggestion

```

**Review by BatuhanW — APPROVED** [review]

**Review by yildirayunlu — APPROVED** [review]

**Pull request merged**